### PR TITLE
core/lib: define MAYBE_UNUSED

### PR DESCRIPTION
**Pull request by kfessel**
There are static functions and variables that may be unused if certain modules are used or not used.
Instead of adding a guard around these marking them as "maybe unused" is much clearer C23 seems to add a stadat attribute system where [[maybe_unused]]  would be that mark

### Contribution description

this adds a`MAYBE_UNUSED` define that uses the badly name `__attribute((unused))__` to mark such `maybe unused` cases

### Testing procedure

read

### Issues/PRs references

this contains #18882

### CODE REVIEW

File: core/lib/include/compiler_hints.h [hunk 1]
```diff
@@ -30,11 +30,13 @@ extern "C" {
  * @brief The *NORETURN* keyword tells the compiler to assume that the function
  *        cannot return.
  */
+#ifndef NORETURN
```
Comment (jue89) [inline core/lib/include/compiler_hints.h:33]: This is unrelated to the commit message "define MAYBE_UNUSED"

File: core/lib/include/compiler_hints.h [hunk 2]
```diff
@@ -43,11 +45,26 @@ extern "C" {
  *        function can be subject to common subexpression elimination and loop
  *        optimization just as an arithmetic operator would be.
  */
+#ifndef PURE
```
Comment (jue89) [inline core/lib/include/compiler_hints.h:48]: Dito.

**Review by jue89 — CHANGES_REQUESTED** [review]

File: core/lib/include/compiler_hints.h (hunk 1 above)
Comment (kfessel) [reply core/lib/include/compiler_hints.h:33]: this contains https://github.com/RIOT-OS/RIOT/pull/18882

Comment (jue89) [reply core/lib/include/compiler_hints.h:33]: Yeah, of course it's useful! But I recommend to split this off into its own commit ;-)

**Comment by jue89** [comment]
You may rebase on master ;-)

**Review by benpicco — APPROVED** [review]
